src/main/osd/osd_elements.c

```diff
@@ -212,30 +212,81 @@ int osdConvertTemperatureToSelectedUnit(int tempInDegreesCelcius)
 static void osdFormatAltitudeString(char * buff, int32_t altitudeCm)
 {
     const int alt = osdGetMetersToSelectedUnit(altitudeCm) / 10;
+    uint8_t roundPart = 0;
 
     int pos = 0;
     buff[pos++] = SYM_ALTITUDE;
     if (alt < 0) {
         buff[pos++] = '-';
     }
-    tfp_sprintf(buff + pos, "%01d.%01d%c", abs(alt) / 10 , abs(alt) % 10, osdGetMetersToSelectedUnitSymbol());
+
+    switch (osdConfig()->altitude_decimals) {
+    case 0:
+        if ((abs(alt) % 10) >= 5) {
+            roundPart = 1;
+        }
+        tfp_sprintf(buff + pos, "%01d%c", abs(alt) / 10 + roundPart, osdGetMetersToSelectedUnitSymbol());
+        break;
+    default:
+        tfp_sprintf(buff + pos, "%01d.%01d%c", abs(alt) / 10, abs(alt) % 10, osdGetMetersToSelectedUnitSymbol());
+        break;
+    }
 }
 
 #ifdef USE_GPS
-static void osdFormatCoordinate(char *buff, char sym, int32_t val)
+static void osdFormatCoordinate(char *buff, char sym, int32_t val, uint8_t coord_dec)
 {
-    // latitude maximum integer width is 3 (-90).
-    // longitude maximum integer width is 4 (-180).
-    // We show 7 decimals, so we need to use 12 characters:
-    // eg: s-180.1234567z   s=symbol, z=zero terminator, decimal separator  between 0 and 1
+    // latitude maximum integer width is 3 (-90)
+    // longitude maximum integer width is 4 (-180)
+    // decimals from 4 to 7 according to user setting
+
+    // example for 7 decimals:
+    // total maximum length is 13 plus zero terminator:
+    // s-180.1234567z
+    // s=symbol, z=zero terminator, decimal separator between 0 and 1
 
     int pos = 0;
     buff[pos++] = sym;
     if (val < 0) {
         buff[pos++] = '-';
         val = -val;
     }
-    tfp_sprintf(buff + pos, "%d.%07d", val / GPS_DEGREES_DIVIDER, val % GPS_DEGREES_DIVIDER);
+
+    char coordFormat[8];
+    int32_t decimalPart = labs(val % GPS_DEGREES_DIVIDER);
+    int32_t roundPart = 0;
+
+    switch (coord_dec) {
+    case 4:
```

Comment:
Expressing this as a case statement is wasteful and prone to errors - much better pre-calculate the divisor based on the number of digits, and then have the code for the split into integer and decimal part only once.

Reply:
I fully agree. At first I thought I could use the usual C wildcards ("*") in the formatting string and pass the argument which unfortunately `tfp_printf` cannot handle. So I had to switch cases anyway for the formatting string, since tbh I've struggled with an attempt at concatenation and `tfp_sprintf` expecting a constant. I will resolve this as soon as the new `printf` implementation you mentioned above [#8813 ](https://github.com/betaflight/betaflight/pull/8813#discussion_r320694155) is in place.